components/movementsensor/wheeledodometry/wheeled_odometry.go

```diff
@@ -0,0 +1,341 @@
+		LinearAccelerationSupported: false,
+	}, nil
+}
+
+func (o *odometry) Close(ctx context.Context) error {
+	o.cancelFunc()
+	o.activeBackgroundWorkers.Wait()
+	return nil
+}
+
+// trackPosition uses the motor positions to calculate an estimation of the position, orientation,
+// linear velocity, and angular velocity of the wheeled base.
+// The estimations in this function are based on the math outlined in this article:
+// https://stuff.mit.edu/afs/athena/course/6/6.186/OldFiles/2005/doc/odomtutorial/odomtutorial.pdf
+func (o *odometry) trackPosition(ctx context.Context) {
+	o.activeBackgroundWorkers.Add(1)
+	utils.PanicCapturingGo(func() {
+		defer o.activeBackgroundWorkers.Done()
+		ticker := time.NewTicker(time.Duration(o.timeIntervalMSecs) * time.Millisecond)
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			default:
+			}
+
+			// Sleep until it's time to update the position again.
+			select {
+			case <-ctx.Done():
+				return
+			case <-ticker.C:
+			}
+
+			// always use the first pair until more than one pair of motors is supported in this model
+			leftMove, err := o.motors[0].left.IsMoving(ctx)
+			if err != nil {
+				o.logger.Error(err)
+				return
+			}
+			rightMove, err := o.motors[0].right.IsMoving(ctx)
+			if err != nil {
+				o.logger.Error(err)
+				return
+			}
+			if !leftMove && !rightMove {
+				// neither motors are moving, so the odometry values do not need to be updated
+				o.linearVelocity.Y = 0
+				o.angularVelocity.Z = 0
+				continue
+			}
+
+			// use GetInParallel to ensure the left and right motors are polled at the same time
+			positionFuncs := func() []rdkutils.FloatFunc {
+				fs := []rdkutils.FloatFunc{}
+
+				o.mu.Lock()
+				defer o.mu.Unlock()
+				// always use the first pair until more than one pair of motors is supported in this model
+				fs = append(fs, func(ctx context.Context) (float64, error) { return o.motors[0].left.Position(ctx, nil) })
+				fs = append(fs, func(ctx context.Context) (float64, error) { return o.motors[0].right.Position(ctx, nil) })
+
+				return fs
+			}
+
+			_, positions, err := rdkutils.GetInParallel(ctx, positionFuncs())
+			if err != nil {
+				o.logger.Error(err)
+				continue
+			}
+
+			// current position of the left and right motors in revolutions
+			if len(positions) != len(o.motors)*2 {
+				o.logger.Error("error getting both motor positions, trying again")
+				continue
+			}
+			left := positions[0]
+			right := positions[1]
+
+			// difference in the left and right motors since the last iteration, in mm
+			leftDist := (left - o.lastLeftPos) * o.wheelCircumference
+			rightDist := (right - o.lastRightPos) * o.wheelCircumference
+
+			// update lastLeftPos and lastRightPos to be the current position in mm
+			o.lastLeftPos = left
+			o.lastRightPos = right
+
+			// linear and angular distance the center point has traveled
+			centerDist := (leftDist + rightDist) / 2
+			centerAngle := (rightDist - leftDist) / o.baseWidth
```

Comment:
I still think we should have a comment about how we're using the small angle assumption in here, and maybe a link to the paper you showed me. At first glance, this looks wrong to me, and I could use some help recognizing that it really is right.

Reply:
I did link the paper above the entire `trackPosition` function because all the math is based off of that. but I can add a comment just to reiterate that it's making an assumption that the time interval is small enough that the angle is small. I don't really think it's necessary to link the whole paper again though